### DISCUSSION TITLE
Device configuration history

### DISCUSSION THREAD

**Discussion by settlej** [netbox-community/netbox#11392] [General]
Is there a way to save (config offline snapshot) the currently pulled configuration and review incase device goes offline and can't use napalm?
Config history should be cached until new update.  Or a gold config.

**Comment by mtinberg** [comment]
As was brought up on the list recently, there are other tools which are a better fit for maintaining config backups and change control, that can use Netbox to provide the inventory of what devices _should_ be backed up, the example given was using an Ansible playbook to provide RANCID-like service, with the list of devices and their status coming from Netbox.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: settlej ***@***.***>
Sent: Wednesday, January 4, 2023 11:59 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Device configuration history (Discussion #11392)


Is there a way to save (config offline snapshot) the currently pulled configuration and review incase device goes offline and can't use napalm?
Config history should be cached until new update. Or a gold config.

**Reply by settlej** [reply]
Then I fail to see the point of the config napalm pull live on the netbox device page.  Its kind of pointless feature if you can't use it for devices that die or unreachable.  Might as well just ssh to the device instead.

**Reply by jeremystretch** [reply]
It's merely a convenience feature. It's also being removed from NetBox v3.5 and ported to a plugin.

**Reply by settlej** [reply]
Jeremy I don't think you should remove it.  Having a configuration is critical for device recovery and powerful when doing scripts to search a single database of running configs.  Otherwise I have to run another dedicated program that I have to maintain when its built into netbox.  Just need the option to save last known config pull at minimum or even better save multiple archives to the postgresql database.

**Reply by dreng** [reply]
> Jeremy I don't think you should remove it. Having a configuration is critical for device recovery and powerful when doing scripts to search a single database of running configs. Otherwise I have to run another dedicated program that I have to maintain when its built into netbox. Just need the option to save last known config pull at minimum or even better save multiple archives to the postgresql database.

I disagree. While the functionality itself seems to be useful, it is not the way NetBox is meant to be. @mtinberg already wrote how to do it properly. As Netbox is a source of truth, it is essential that NetBox doesn't accept another source of truth. Pulling and storing a config from any device into NetBox is the opposite of being a source of truth. If you still want to use it like this, you have the ability to use napalm as a plugin after it has been ported. That's not a big deal. You just need to add an entry to the config and another one to the requirements text file. After that the plugin upgrades automatically every time you update NetBox itself.

You should see the division into several tools not so much as a burden, but as freedom.

**Comment by mtinberg** [comment]
I'm don't know for sure, but I would guess that the (optional) NAPALM support is for the Environmental display and LLDP to Interface Description audit that's built-in and it's left in as sometimes viewing the config is useful without needing to reference another tool, or having audit Reports or Plugins within Netbox able to pull the config (NAPALM is available over the API too).  There are some cases where it's useful, so it exists, but it's optional and for more comprehensive config management tools you are going to want something else, Netbox is mainly about having a database of how things are supposed to be to be able to compare actual configs against.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Comment by mtinberg** [comment]
I think there is a small misunderstanding here, moving the functionality to a plugin doesn't make the functionality disappear, it just moves where  the code is hosted, so you'd need to install the plugin when upgrading and the functionality in your netbox instance would be the same.  This has always been an optional feature, but plugins didn't exist when it was originally developed, so now that plugins exist it made more sense to re-organize the code as a plugin, like all other optional features.

I think there is probably a benefit to you in that developing a plugin has a lower barrier to change than core netbox where changes impact every user and have to be carefully managed for long-term impact to project maintenance, changes to plugins only affect people who use them and the user community around the plugin can decide for themselves what is the most useful behavior, so you have more chance to develop your RANCID-like feature in a the plugin than in core.
________________________________

**Comment by jcralbino** [comment]
If you want a configuration manager check rConfig

I agree that while it will be nice to have on ui for those things in netbox the truth is that are other tools better for configuration backup.

Using the configuration manager API you can create some reports for devices in netbox

**Comment by candlerb** [comment]
I'd suggest using [oxidized](https://github.com/ytti/oxidized).  It retrieves configs over ssh or telnet, stores them in a git repo, has a web interface for browsing history, and optionally can send out E-mail notifications whenever a config changes.

Generating the oxidized inventory (router.db) from Netbox should be relatively straightforward, and you could make a custom link to `https://oxidized.example.net/node/show/<nodename>`

Oxidized is fully free and open source, although the maintainer is looking for someone else to maintain it, and there hasn't been a release since 2020 so you're best off installing from git head.

I haven't used rConfig, but I note that its website is broken: there are important FAQ links like "Why is rConfig v6 not free?" which don't open (at least, not in Chrome)

**Reply by jcralbino** [reply]
You have a free version of rConfig although the paid version is quite acceptable in terms of price 

https://github.com/rconfig/rconfig?utm_source=newsletter&utm_medium=email&utm_campaign=Huge+rConfig+announcements

**Reply by candlerb** [reply]
That's a link to their github repository, and I see the licence file says GPLv3.

So why does their homepage say "Free to download and use V3.x ONLY"?

Why does the same page have a FAQ entry for "Why is rConfig v6 not free?"

If you inspect the HTML source (to get past their broken Javascript), the answer to that question is:

> We have spent nearly 18 months developing rConfig v6. It is a professional, high end and quality assured network management tool. We know that businesses that need this tool are happy to pay to sustain and deliver great features for rConfig in the future.
>                
> Also, rConfig has a high ROI, and very low TCO. Compared against other NCMs in the market, we have an unlimited device licensing model (others charge for the number of devices under management) and no hidden costs such as Windows or SQL license fees.

Maybe the software is OK - but the website does not inspire confidence.
